.github/workflows/release.yml

```diff
@@ -5,82 +5,177 @@ permissions:
+        default: ''
+env:
+  USE_COMMIT: ${{ github.event.inputs.commit }}
+  FORCE_RELEASE: ${{ github.event.inputs.force }}
 
 jobs:
+  setup_job:
+    name: Setup workflow
+    runs-on: ubuntu-latest
+    outputs:
+      version: ${{ steps.version_step.outputs.VERSION }}
+      target_sha: ${{ steps.version_step.outputs.GIT_SHA }}
+      latest_tag: ${{ steps.docker_step.outputs.LATEST_TAG }}
+      release_tag: ${{ steps.docker_step.outputs.RELEASE_TAG }}
+    steps:
+      - name: Set target sha and Fluvio version
+        id: version_step
+        run: |
+          if [[ -z "${USE_COMMIT}" ]]; then
+            GITHUB_VERSION=$(curl -fsS https://raw.githubusercontent.com/infinyon/fluvio/${{ github.sha }}/VERSION)
+            echo "VERSION=${GITHUB_VERSION}" >> $GITHUB_ENV
+            echo "::set-output name=VERSION::${GITHUB_VERSION}
+            echo "GIT_SHA=${{ github.sha }}" >> $GITHUB_ENV
+            echo "::set-output name=GIT_SHA::${{ github.sha }}"
+          else
+            GITHUB_VERSION=$(curl -fsS https://raw.githubusercontent.com/infinyon/fluvio/${{ github.event.inputs.commit }}/VERSION)
+            echo "VERSION=${GITHUB_VERSION}" >> $GITHUB_ENV
+            echo "::set-output name=VERSION::${GITHUB_VERSION}"
+            echo "GIT_SHA=${{ github.event.inputs.commit }}" >> $GITHUB_ENV
+            echo "::set-output name=GIT_SHA::${{ github.event.inputs.commit }}"
+          fi
+      - name: Set Docker tag related env vars
+        id: docker_step
+        run: |
+            echo "::set-output name=LATEST_TAG::infinyon/fluvio:${{ env.VERSION }}-${{ env.GIT_SHA }}"
+            echo "::set-output name=RELEASE_TAG::infinyon/fluvio:${{ env.VERSION }}"
+
+  # Check for Github Release
   release_github:
-    name: Create GitHub Release
+    name: Release Fluvio to GitHub Release
+    needs: [setup_job]
     runs-on: ubuntu-latest
     permissions: write-all
     steps:
-      - uses: actions/checkout@v2
       - name: Login GH CLI
         run: gh auth login --with-token < <(echo ${{ secrets.GITHUB_TOKEN }})
+      - name: Attempt to download release by version number 
+        id: release_check
+        continue-on-error: true
+        run: |
+          gh release download -R infinyon/fluvio "v${{ needs.setup_job.outputs.version }}"
+      # if the check fails, then continue with release
       - name: Download artifacts from dev
+        if: ${{ steps.release_check.outcome == 'failure' }}
         run: gh release download -R infinyon/fluvio dev
       - name: Create GH Release
+        if: ${{ steps.release_check.outcome == 'failure' }}
         run: |
-          export VERSION="$(cat VERSION)"
-          gh release create -R infinyon/fluvio --title="v${VERSION}" -F ./CHANGELOG.md "v${VERSION}" \
+          gh release create -R infinyon/fluvio --title="v${{ needs.setup_job.outputs.version }}" -F ./CHANGELOG.md "v${{ needs.setup_job.outputs.version }}" \
```

Comment:
Can you assign this to a variable once and use that variable? e.g.

```
export VERSION="${{ needs.setup_job.outputs.version }}"
...
```

Reply:
That's a good idea to clean up. I carried the pattern over to all the separate jobs.